packages/x-components/src/x-modules/url/components/url-handler.vue

```diff
@@ -99,28 +113,102 @@
+      const metadata = this.createWireMetadata();
+      this.$x.emit('ParamsLoadedFromUrl', all, metadata);
+      this.$x.emit('ExtraParamsLoadedFromUrl', extra, metadata);
       // TODO: Move this logic from here.
       if (all.query) {
-        this.$x.emit('UserOpenXProgrammatically');
+        this.$x.emit('UserOpenXProgrammatically', undefined, metadata);
       }
       this.urlLoaded = true;
     }
 
+    /**
+     * Creates the wire metadata to include in every emitted {@link XEvent | XEvents}.
+     *
+     * @returns The {@link WireMetadata | metadata}.
+     * @internal
+     */
+    protected createWireMetadata(): Pick<WireMetadata, 'feature' | 'location'> {
+      return {
+        feature: 'url',
+        location: this.detectLocation()
+      };
+    }
+
+    /**
+     * Detects the {@link FeatureLocation | location} used to build the
+     * {@link QueryOriginInit | events metadata origin}.
+     *
+     * @returns The {@link FeatureLocation | location}.
+     * @internal
+     */
+    protected detectLocation(): FeatureLocation {
+      const currentUrl = new URL(window.location.href);
+      const previousUrl = this.url;
+      this.url = currentUrl;
+
+      const isInternalNavigation =
+        previousUrl?.search !== currentUrl.search && previousUrl?.pathname === currentUrl.pathname;
+      if (isInternalNavigation) {
+        return 'url_history';
+      }
+
+      const isNavigatingFromPDP = this.isNavigatingFromPdp();
+      if (isNavigatingFromPDP) {
+        return 'url_history_pdp';
+      }
+
+      return 'external';
+    }
+
+    /**
+     * Check if the navigation is from a product page.
+     *
+     * @remarks Due to Safari 14 not supporting the new and standard PerformanceNavigationTiming
+     * API, we are falling back to the deprecated one, PerformanceNavigation. We also fallback to
+     * this API whenever we get a navigationType equal to reload, because Safari has a bug that the
+     * navigationType is permanently set to reload after you have reload the page and it never
+     * resets.
+     *
+     * @returns True if the navigation is from a product page, false otherwise.
+     * @internal
+     */
+    protected isNavigatingFromPdp(): boolean {
+      const navigationEntries = window.performance.getEntriesByType('navigation');
+      const navigationType = (navigationEntries[0] as PerformanceNavigationTiming)?.type;
+      const useFallbackStrategy =
+        window.performance.navigation &&
+        (isArrayEmpty(navigationEntries) || navigationType === 'reload');
+
+      if (useFallbackStrategy) {
+        const {
+          type: fallbackNavigationType,
+          TYPE_BACK_FORWARD,
+          TYPE_NAVIGATE
+        } = window.performance.navigation;
+        const isNavigatingInSPA =
+          this.snippetConfig?.isSPA === false && fallbackNavigationType === TYPE_NAVIGATE;
+        return fallbackNavigationType === TYPE_BACK_FORWARD || isNavigatingInSPA;
+      } else {
```

Comment:
I prefer to avoid the use of else if it's possible and in this case it's as you're returning inside the if.